homeassistant/components/snapcast/media_player.py

```diff
@@ -114,195 +119,197 @@ class SnapcastGroupDevice(MediaPlayerEntity):
         | MediaPlayerEntityFeature.SELECT_SOURCE
     )
 
-    def __init__(self, group, uid_part, entry_id):
-        """Initialize the Snapcast group device."""
+    def __init__(
+        self,
+        device: Snapgroup | Snapclient,
+        entry_id: str,
+    ) -> None:
+        """Initialize the base device."""
         self._attr_available = True
-        self._group = group
+        self._device = device
         self._entry_id = entry_id
-        self._attr_unique_id = f"{GROUP_PREFIX}{uid_part}_{self._group.identifier}"
+
+    def _append(self) -> None:
+        """Add self to appropriate list."""
+        raise NotImplementedError
+
+    def _remove(self) -> None:
+        """Remove self from appropriate list."""
+        raise NotImplementedError
+
+    @property
+    def current_group(self) -> Snapgroup:
+        """Return the group."""
+        raise NotImplementedError
```

Comment:
So why do we add those?

Reply:
The `current_group` property allows a common definition for `source`, `source_list` and `async_select_source` in the base class. In retrospect, it should probably be renamed to `_current_group` as it is only accessed from within the classes.

Similarly `_append` and `_remove` allow common definitions for `async_added_to_hass` and `async_will_remove_from_hass` in the base class. I had felt that raising `NotImplementedError` was preferred to the default `AttributeError` in the case a developer forgets to implement the methods in the derived class.

Reply:
Is there a way we can refactor this? We don't want to reference Entities directly after they're created so we should not store them in such list.

Reply:
Ok. What is the preferred way to reference an entity after it is created?

Reply:
Not :)

Using entity services, which call a function on the entity (see snapshot and that other one). 

Other services usually directly reference the client and use that to send requests to the device or service, but i dont think that's a good fit here

Reply:
I see. The lists are primarily used to determine if a client/group entity already exists. Since clients/groups can come and go from the Snapcast server we need to be able to determine if an entity has already been created.

Reply:
Where is the logic where we add an entity?

Reply:
We can create a list of identifiers for which we created entities. Anytime we want to add one, we can check that list

Reply:
Ideally we move the code that is required for adding entities to the platform and keep it there. Check AirGradient for an example (withings also is a good one, but it's bigger)

Reply:
In the current implementation each entity adds itself to the corresponding list, e.g. `self.hass.data[DOMAIN][self._entry_id].groups` in `async_added_to_hass()`.
If there is an update from the server (only a single callback exists) this lists are used to find out which entities need to be added/removed.
> We can create a list of identifiers for which we created entities. Anytime we want to add one, we can check that list

This works for adding entities, but how can we remove entities based on the identifier?

Reply:
Do the same thing, compare with your current list and if you have leftovers remove them. Checkout tedee for an example

Reply:
Ok, so it appears the refactor would probably be:
A. Refactor the `HomeAssistantSnapcast` class into a proper update coordinator derived from `DataUpdateCoordinator`.
B. Push entity addition/removal into `async_setup_entry` by registering a listener with the coordinator.

Reply:
@luar123 I think the removal would work by looking up the entities in the registry via their identifier.

Reply:
Ok, that ended up being a lot more work than I anticipated, but I think I've got it,